### PR TITLE
COL-74  - Implement Manage categories REST APIs

### PR DESCRIPTION
**Pull request by nicolaasmatthijs**
Implement Manage categories REST APIs

https://jira.ets.berkeley.edu/jira/browse/COL-74

### CODE REVIEW

**Comment by simong** [comment]
When starting the application I'm getting:

```
$ node app
[2015-04-02T08:14:07.110Z]  INFO: col-core/server/8962 on dhcp-172-17-169-250.eduroam.wireless.private.cam.ac.uk: Collabosphere server is listening at http://127.0.0.1:2000
(node) warning: possible EventEmitter memory leak detected. 11 listeners added. Use emitter.setMaxListeners() to increase limit.
Trace
    at WriteStream.addListener (events.js:160:15)
    at WriteStream.Readable.on (_stream_readable.js:707:33)
    at PrettyStream.Stream.pipe (stream.js:65:8)
    at createLogger (/Users/sg555/projects/fronteer/berkeley/collabosphere/node_modules/col-core/lib/logger.js:63:18)
    at module.exports (/Users/sg555/projects/fronteer/berkeley/collabosphere/node_modules/col-core/lib/logger.js:42:21)
    at Object.<anonymous> (/Users/sg555/projects/fronteer/berkeley/collabosphere/node_modules/col-lti/lib/api.js:23:41)
    at Module._compile (module.js:456:26)
    at Object.Module._extensions..js (module.js:474:10)
    at Module.load (module.js:356:32)
    at Function.Module._load (module.js:312:12)
(node) warning: possible EventEmitter memory leak detected. 11 listeners added. Use emitter.setMaxListeners() to increase limit.
Trace
    at WriteStream.addListener (events.js:160:15)
    at WriteStream.Readable.on (_stream_readable.js:707:33)
    at PrettyStream.Stream.pipe (stream.js:99:8)
    at createLogger (/Users/sg555/projects/fronteer/berkeley/collabosphere/node_modules/col-core/lib/logger.js:63:18)
    at module.exports (/Users/sg555/projects/fronteer/berkeley/collabosphere/node_modules/col-core/lib/logger.js:42:21)
    at Object.<anonymous> (/Users/sg555/projects/fronteer/berkeley/collabosphere/node_modules/col-lti/lib/api.js:23:41)
    at Module._compile (module.js:456:26)
    at Object.Module._extensions..js (module.js:474:10)
    at Module.load (module.js:356:32)
    at Function.Module._load (module.js:312:12)
(node) warning: possible EventEmitter memory leak detected. 11 listeners added. Use emitter.setMaxListeners() to increase limit.
Trace
    at WriteStream.addListener (events.js:160:15)
    at WriteStream.Readable.on (_stream_readable.js:707:33)
    at PrettyStream.Stream.pipe (stream.js:121:8)
    at createLogger (/Users/sg555/projects/fronteer/berkeley/collabosphere/node_modules/col-core/lib/logger.js:63:18)
    at module.exports (/Users/sg555/projects/fronteer/berkeley/collabosphere/node_modules/col-core/lib/logger.js:42:21)
    at Object.<anonymous> (/Users/sg555/projects/fronteer/berkeley/collabosphere/node_modules/col-lti/lib/api.js:23:41)
    at Module._compile (module.js:456:26)
    at Object.Module._extensions..js (module.js:474:10)
    at Module.load (module.js:356:32)
    at Function.Module._load (module.js:312:12)
[2015-04-02T08:14:07.218Z]  INFO: col-core/server/8962 on dhcp-172-17-169-250.eduroam.wireless.private.cam.ac.uk: Adding /lti/ to list of paths that are not CSRF-protected.
[2015-04-02T08:14:07.220Z]  INFO: col-core/8962 on dhcp-172-17-169-250.eduroam.wireless.private.cam.ac.uk: Finished initializing REST APIs
[2015-04-02T08:14:07.220Z]  INFO: app/8962 on dhcp-172-17-169-250.eduroam.wireless.private.cam.ac.uk: Collabosphere has started up. Enjoy!
```

File: node_modules/col-categories/lib/api.js [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+ *
+ * @param  {Context}        ctx                         Standard context containing the current user and the current course
+ * @param  {Function}       callback                    Standard callback function
+ * @param  {Object}         callback.err                An error that occurred, if any
+ * @param  {Category[]}     callback.categories         The categories in the current course
+ */
+var getCategories = module.exports.getCategories = function(ctx, callback) {
+  // Get the categories from the DB. As we need to include the number of assets per category,
+  // a raw query is used
+  var sqlQuery = 'SELECT category.*, (SELECT COUNT(*)::int FROM assets_categories WHERE category_id = category.id) AS asset_count';
```
Comment (simong) [inline node_modules/col-categories/lib/api.js]: Although I doubt it'll ever matter and it depends on things like indexes and the set of data in the tables, doing a subselect to get the count might be suboptimal.

I would prefer something such as:
```
SELECT category.*, count(assets_categories.asset_id) AS asset_count
FROM categories AS category 
JOIN assets_categories ON assets_categories.category_id = category.id
WHERE category.course_id = 1
GROUP BY category.id;
```

File: node_modules/col-core/lib/db.js [hunk 2]
```diff
@@ -457,12 +466,17 @@ var setUpModel = function(sequelize) {
...
+   * @property {String}       title                   The name of the category
+   * @property {Number}       [canvas_assignment_id]  The id of the assignment to which this category is linked
    */
   var Category = module.exports.Category = sequelize.define('category', {
-    'display_name': {
+    'title': {
       'type': Sequelize.STRING,
       'allowNull': false
+    },
+    'canvas_assignment_id': {
```
Comment (simong) [inline node_modules/col-core/lib/db.js]: Do categories have to be scoped to assigments? I thought you would define them per course?

File: node_modules/col-core/lib/util.js [hunk 3]
```diff
@@ -53,3 +55,24 @@ var getBooleanParam = module.exports.getBooleanParam = function(value, defaultVa
...
+ * @param  {Object}             value           The value to wrap in an array
+ * @return {Object[]}                           The provided value wrapped in an arrray
+ * @see http://underscorejs.org/#toArray
+ */
+var toArray = module.exports.toArray = function(value) {
+  if (!value) {
+    return [];
+  }
+
+  // Underscore doesn't wrap primitive values
```
Comment (simong) [inline node_modules/col-core/lib/util.js]: Note that we're using `lodash` and not `underscore` (same note for the @see).

File: node_modules/col-categories/lib/api.js [hunk 4]
```diff
@@ -0,0 +1,292 @@
...
+ * @param  {Number}         [canvasAssignmentId]        The id of the assignment to which this category is linked
+ * @param  {Function}       callback                    Standard callback function
+ * @param  {Object}         callback.err                An error that occurred, if any
+ * @param  {Category}       callback.category           The created category
+ */
+var createCategory = module.exports.createCategory = function(ctx, title, canvasAssignmentId, callback) {
+  // Parameter validation
+  var validationSchema = Joi.object().keys({
+    'title': Joi.string().max(255).required(),
+    'canvasAssignmentId': Joi.number().optional()
```
Comment (simong) [inline node_modules/col-categories/lib/api.js]: This assignment id doesn't seem to be used anywhere. Do we need it?

**Comment by simong** [comment]
PR LGTM, left a couple of questions

File: node_modules/col-core/lib/db.js (hunk 2 above)
Comment (nicolaasmatthijs) [inline node_modules/col-core/lib/db.js]: Categories are optionally associated to an assignment in the course. When a new assignment is created in the assignments tool, this should be added as a category in the asset library. These categories should be highlighted as "special" categories in the UI.

File: node_modules/col-categories/lib/api.js (hunk 4 above)
Comment (nicolaasmatthijs) [inline node_modules/col-categories/lib/api.js]: It's not possible to set it through the REST APIs, but once we'll need it to create a new category when a new assignment is set up.

**Comment by nicolaasmatthijs** [comment]
Followed up on review. Back to @simong

File: node_modules/col-assets/tests/test-categories.js [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+          CategoriesTestUtil.assertCreateCategory(client, course, 'Category 1', function(category1) {
+            CategoriesTestUtil.assertCreateCategory(client, course, 'Category 2', function(category2) {
+
+              var opts = {'categories': [category1.id, category2.id]};
+              AssetsTestUtil.assertCreateLink(client, course, 'UC Berkeley', 'http://www.berkeley.edu/', opts, function(asset) {
+                CategoriesTestUtil.assertCategory(asset.categories[0], category1);
+                CategoriesTestUtil.assertCategory(asset.categories[1], category2);
+
+                // Ensure that the categories are returned when getting the asset
+                AssetsTestUtil.assertGetAsset(client, course, asset.id, asset, 0, function(asset) {
```
Comment (simong) [inline node_modules/col-assets/tests/test-categories.js]: This asserts that the returned categories are well-formed, it doesn't assert that the correct categories were returned

File: node_modules/col-assets/tests/util.js [hunk 6]
```diff
@@ -53,6 +55,14 @@ var assertAsset = module.exports.assertAsset = function(asset, expectedAsset, ex
   assert.ok(!asset.user.points);
   assert.ok(!asset.user.share_points);
 
+  // Ensure that valid categories are present
+  if (expectedCategories) {
+    assert.ok(_.isArray(asset.categories));
+    _.each(asset.categories, function(category) {
+      CategoriesTestUtil.assertCategory(category);
+    });
+  };
```
Comment (simong) [inline node_modules/col-assets/tests/util.js]: Unneeded `;`

File: node_modules/col-assets/tests/test-categories.js (hunk 5 above)
Comment (nicolaasmatthijs) [inline node_modules/col-assets/tests/test-categories.js]: It does actually. When creating the link, we assert that both categories are present. When getting the link, we pass in `asset` as the expected asset, which will compare the categories as well. I can add an explicit check here if you want though.

File: node_modules/col-assets/tests/util.js [hunk 7]
```diff
@@ -100,6 +110,16 @@ var assertAsset = module.exports.assertAsset = function(asset, expectedAsset, ex
       assert.strictEqual(asset.views, expectedAsset.views);
     }
 
+    // Ensure that the expected categories are present
+    if (expectedCategories) {
```
Comment (simong) [inline node_modules/col-assets/tests/util.js]: there are 2 `if (expectedCategories) {` blocks in here, which tripped me up. Can those be merged?

Comment (nicolaasmatthijs) [inline node_modules/col-assets/tests/util.js]: It wouldn't be in line with the rest of the function. The first if checks the provided asset itself, whilst the second if verifies it against the expected asset.

Comment (simong) [inline node_modules/col-assets/tests/util.js]: 2 spaces indenting makes everything more complicated than it should be -_-

**Comment by simong** [comment]
Followed-up and merged in

**Pull request merged**